### ISSUE TITLE
Adding a meta section to a private thread

### ISSUE THREAD

**Issue by chriseman** [chat-sdk/chat-sdk-ios#198]
Hi,

We need to add additional data e.g. **chat requested accepted** to each thread, the only way I figured out to do so, at least this works for the user object, is the meta tag. The PThread object provides in it's structure a variable named meta.
So I tried to add a **meta** section to my private test thread in the Firebase database. However I haven't found a way to read the meta data on the iOS client side yet. There are two methods _setMetaString(value: forKey:)_ and _metaString(forKey:)_ but I suspect these two methods only add meta information locally on the device, and not in the Firebase database.

<img width="236" alt="screen shot 2018-01-25 at 12 12 18" src="https://user-images.githubusercontent.com/14274762/35385596-17501f88-01c9-11e8-8a50-3da9c6fe512a.png">

That's the data I receive when I print the private thread object on the console

<img width="804" alt="screen shot 2018-01-25 at 12 13 32" src="https://user-images.githubusercontent.com/14274762/35385645-4a27e3fa-01c9-11e8-905f-dae8da9aa46a.png">

Am I missing here anything?

**UPDATE**
I also added the meta section to the **details** section, same result. 

Thanks,
Chris

**Comment by simonsmileyandrews** [comment]
Hi @chriseman,

When adding meta data to user objects it is actually maintained across Firebase and the device. You can see we use this to update the user object:

```
    NSString * email = provider.email;
    if (email && !_model.email) {
        [_model setMetaString:email forKey:bEmailKey];
    } 
```

this is then updated on the user object in the user meta:

<img width="376" alt="screen shot 2018-02-05 at 3 42 24 pm" src="https://user-images.githubusercontent.com/33849506/35788751-71ff93c6-0a8b-11e8-89a9-9e82c0007394.png">

The way to architect and code this really depends on the exactly functionality you are after. It also might require some additional meta data coding. 

**Functionality:**
Depending on your functionality will depend on how it should be coded. For example if you want to implement a request system you need to add quite a bit of additional code. 

You will need to have an "invited users" section when the thread is created. The thread would then need a status (created, sent, accepted, rejected) and this would then need to be set for each user. This would mean you would need users -> status -> userStatus in the meta data. 

You will also need new views which show user's invited chats and also allow them to accept or reject them. It becomes more complicated with multiple group chats as you need to decide what happens if one user rejects but others accept, you'll need to remove the rejected users from the chat before allowing others to chat.

**Execution:** 
One of the issues you might be facing is that the setMetaString is a function for user objects. You are working with thread objects meaning that it won't work. This is probably why you haven't seen it updating on Firebase. It would be worth you playing around with a user object and seeing how the meta data updates when set on a user object. 

If you want this functionality for thread you will probably need to write a lot of it yourself. The good thing is that you can work closely with the User object as the functionality is almost identical. 

To advise on this it would be useful to know more details on the functionality you are trying to achieve. We have previously adapted the ChatSDK with event invitations and so chat invitations would definitely be possible.

Simon

**Comment by bensmiley** [comment]
I've just added an API to do this:

```
CCThreadWrapper * thread = [CCThreadWrapper threadWithModel:threadModel];

[thread.model setMetaValue:@"Test" forKey:@"value"];
[thread metaOn];

[[NSNotificationCenter defaultCenter] addObserverForName:bNotificationThreadMetaUpdated object:Nil queue:0 usingBlock:^(NSNotification * notification) {
    NSLog(@"Updated");
}];
        
[thread pushMeta].thenOnMain(^id(id result) {
    NSLog(@"Success");
    return Nil;
}, ^id(NSError * error) {
    NSLog(@"Error");
    return Nil;
});
```

To use this code you will need to import the `ChatFirebaseAdapter.h` header file. These changes are available in [4.4.0](https://github.com/chat-sdk/chat-sdk-ios/releases/tag/4.4.0). Run `pod update`.
